Refresh the AWS app access guide (#37857)
gravitational/teleport: docs/pages/application-access/cloud-apis/aws-console.mdx

   to your role:

   ```code
   $ ARN=$(aws iam list-policies --output text --query "Policies[?PolicyName=='ReadOnlyAccess'].Arn")
   ```

1. Attach the `ReadOnlyAccess` policy to the role:

   ```code
   $ aws iam attach-role-policy --role-name ExampleReadOnlyAccess --policy-arn $ARN
   ```

### Give the Teleport Application Service permissions to assume roles

In this section, you will create an IAM role that allows the Teleport
Application Service to assume other IAM roles in order to proxy user traffic to
AWS APIs.

1. Define a trust policy to enable the Teleport Application Service to assume
   the role you will create. 

   <Tabs>
   <TabItem label="EC2">

   Create a file called `app-service-tp.json`:

   ```json
   {
     "Version": "2012-10-17",
     "Statement": [
       {
         "Effect": "Allow", 
         "Principal": {
           "Service": "ec2.amazonaws.com"
         },
         "Action": "sts:AssumeRole"
       }
     ]
   }
   ```

   </TabItem>
   <TabItem label="EKS">

   Retrieve your OIDC issuer ID, assigning <Var name="cluster-name" /> to the name
   of your EKS cluster and <Var name="eks-region" /> to the AWS region where your

>    EKS cluster is running:

Author: Should we recommend running `eksctl create iamserviceaccount` with the `--override-existing-serviceaccounts` flag and a `--name` flag pointing to the `teleport-kube-agent` service account? 

One consideration is that, in this guide, we tell the user to install the `teleport-kube-agent` Helm chart (which creates a service account for the Teleport Application Service) after creating the IAM role.

>    EKS cluster is running:

Reviewer: > One consideration is that, in this guide, we tell the user to install the teleport-kube-agent Helm chart (which creates a service account for the Teleport Application Service) after creating the IAM role.

I see. If that's the case then it's good. 

To be honest I don't know what ppl do for IAM role for service account in the field. Personally I would still use eksctl since it's easier, then use that created service account to create App service with Helm.

Maybe @hugoShaka @anto have more insight on this.

  will print its ID.

## Step 1/4. Configure AWS IAM

In this section, you will configure AWS IAM resources to allow the Teleport
Application Service to proxy AWS APIs.

The rest of the guide assumes that you are protecting access to an IAM role
called `ExampleReadOnlyAccess`, which allows read-only access to AWS resources
through the AWS-managed `ReadOnlyAccess` policy. We recommend following this
guide in a demo environment with the `ExampleReadOnlyAccess` role, and
protecting production AWS roles when you are familiar with the procedure.

You will create the following resources:

|Name | Resource | Function                                   |
|---|----------|--------------------------------------------|
|`ExampleReadOnlyAccess`| IAM role | Example role to protect access to with Teleport. |
|`TeleportAWSAccess`|IAM role|Allows the Application Service to assume other roles in order to proxy user requests to AWS.|
|`AssumeRole`|IAM policy|Allows the Application Service to assume other roles in order to proxy user requests to AWS.|
|`TeleportAWSAccess` (for EC2 deployments) |EC2 instance profile|Associates the `TeleportAWSAccess` role with your EC2 instance.|

> ### Configure a role to enable access to

Author: I've suggested "Configure a role for Teleport users to request", which I think makes more sense for this setup step. Otherwise, this section seems to overlap with the later section called, "Give the Teleport Application Service permissions to assume roles".

In this section, you will configure an AWS role to enable access to with the

> Teleport Application Service.

Author: I agree that this sentence is unclear, but I'm not sure this rephrasing helps, since it sounds like only the Application Service requires this access. I've suggested adding more context to this paragraph.

1. Obtain AWS credentials for the account where you will run the Teleport
   Application Service and make them available to your terminal shell.

1. Create a trust policy document, which authorizes an entity to assume the role
   you want to protect access to. To do so, create a file called
   `ro-access.json` with the following content, replacing <Var
   name="AWS_ACCESS_ACCOUNT" /> with the ID of the AWS account where you will
   run the Teleport Application Service:

   ```json
   {
     "Version": "2012-10-17",
     "Statement": [
       {
         "Effect": "Allow",
         "Principal": {
           "AWS": "arn:aws:iam::<Var name="AWS_ACCESS_ACCOUNT" />:role/TeleportAWSAccess"
         },
         "Action": "sts:AssumeRole"
       }
     ]
   }
   ```

>    In the configuration we show in this guide, the Teleport Application Service

Author: "Configuration" should be "setup", referring to everything the reader will have accomplished by the time they finish the guide.

   assumes the `TeleportAWSAccess` role, then uses that role to assume the
   `ExampleReadOnlyAccess` role. With the trust policy above, AWS authorizes
   this operation. (We will create the `TeleportAWSAccess` role later in Step

>    1.)

Author: The role creation happens in the section, "Give the Teleport Application Service permissions to assume roles", which is later in step 1. If this is causing confusion, though, I'll change the wording to "later in this guide".

   ```code
   $ aws iam add-role-to-instance-profile \
   --instance-profile-name TeleportAWSAccess \
   --role-name TeleportAWSAccess
   ```

1. Obtain the ID of the EC2 instance where you will install the Teleport
   Application Service.

1. Run the following command to associate the new instance profile with your
   instance:

   ```code
   $ aws ec2 associate-iam-instance-profile --iam-instance-profile Name="TeleportAWSAccess" \
   --instance-id <Var name="INSTANCE_ID" /> --region <Var name="AWS_REGION" />
   ```

</TabItem>
<TabItem label="EKS">

   Annotate the service account used by the Teleport Application Service to
   instruct it to assume the `TeleportAWSAccess` role:

   ```code
   $ kubectl annotate serviceaccount \
   -n teleport-agent \
   teleport-kube-agent eks.amazonaws.com/role-arn=arn:aws:iam::<Var name="EKS_ACCOUNT" />:role/TeleportAWSAccess
   ```

</TabItem>
</Tabs>

## Step 2/4. Configure Teleport IAM role mapping

In this step, you will define a Teleport role that confers access to the
`ExampleReadOnlyAccess` IAM role you created in the previous step.

1. Create a file called `aws-ro-access.yaml` with the following content,

>    replacing <Var name="AWS_ACCOUNT" /> with the account number where you

Reviewer 2: we refer it as ID elswhere
```suggestion
   replacing <Var name="AWS_ACCOUNT" /> with the account ID where you
```

### Configure the Teleport Application Service

1. On the host where you will run the Teleport Application Service, create a
   file at `/etc/teleport.yaml` with the following content:

   ```yaml
   version: v3
   teleport:
     join_params:
       token_name: "/tmp/token"
       method: token
     proxy_server: "teleport.example.com:443"
   auth_service:
     enabled: off
   proxy_service:
     enabled: off
   ssh_service:
     enabled: off
   app_service:
     enabled: "yes"
     apps:
     - name: "awsconsole"
     # The public AWS Console is used after authenticating the user from Teleport
       uri: "https://console.aws.amazon.com/ec2/v2/home"
   ```

> 1. Edit `/etc/teleport.yaml` to replace `teleport.example.com:443` with the host

Reviewer 2: can we use the same `in place editor` thingy we do elsewhere in this guide for writing over `teleport.example.com:443`?

   and port of your Teleport Proxy Service or Teleport Cloud tenant, e.g.,
   `mytenant.teleport.sh:443`.

   The `app_service` field configures the Teleport Application Service. Each
   item within `app_service.apps` is an application configuration. 

</TabItem>
<TabItem label="EKS">

1. Retrieve the join token you created earlier in this guide by running the

>    following command and copying a token with the `App` type:

Reviewer 2: ```suggestion
   following command and copy the token with the `App` type:
```

already started, and you can skip to [Step 4](#step-44-access-aws-resources).

1. (!docs/pages/includes/aws-credentials.mdx service="the Application Service"!)
1. (!docs/pages/includes/start-teleport.mdx service="the Application Service"!)

<Admonition type="warning" title="non-standard AWS regions">
For non-standard AWS regions such as AWS GovCloud (US) regions and AWS China
regions, set the corresponding region in the `AWS_REGION` environment variable
or in the AWS credentials file so that the Application Service can use the
correct STS endpoint.
</Admonition>

## Step 4/4. Access AWS resources

Now that you have configured the Teleport Application Service to proxy requests
to AWS, users can access AWS resources through Teleport.

### Access the AWS console

1. Visit the home page of the Teleport Web UI or click the **Resources** tab. If
   the Teleport Application Service is proxying the AWS Management Console as
   expected, the Web UI will display the name of the application you registered,
   which this guide assumes is `awsconsole`. (Enter the name of the application
   into the search box if there are too many resources to view all of them at
   once.)

> 1. Click the **Launch** button for the AWS Console application, then choose the

Reviewer 2: ```suggestion
1. Click the **Launch** button for the AWS Console application, then click on the
```

   role you would like to assume when signing in to the AWS Console:

   ![IAM role selector](../../../img/application-access/iam-role-selector.png)

> 1. Click on the role you want to assume and you will get redirected to the AWS

Reviewer 2: didn't i already click on the role on the previous step? 
```suggestion
1. You will get redirected to the AWS
```

   Management Console, signed in with the selected role. You should see your
   Teleport user name as a federated login assigned to `ExampleReadOnlyRole` in
   the top-right corner of the AWS Console:

   ![Federated login](../../../img/application-access/federated-login@2x.png)

### Access the AWS CLI

> 1. Log into the previously configured AWS console app on your desktop:

Reviewer 2: is the `previously` necessary? (b/c i didn't know what was previously configured)
```suggestion
1. Log into the configured AWS console app on your desktop:
```

   ```code
   $ tsh apps login --aws-role ExampleReadOnlyAccess awsconsole
   Logged into AWS app "awsconsole".
   
   Your IAM role:
     arn:aws:iam::000000000000:role/ExampleReadOnlyAccess
   
   Example AWS CLI command:
     tsh aws s3 ls
   
   Or start a local proxy:
     tsh proxy aws --app awsconsole
   ```

   The `--aws-role` flag allows you to specify the AWS IAM role to assume when
   accessing AWS APIs. You can either provide a role name like `--aws-role
   ExampleReadOnlyAccess` or a full role ARN like
   `arn:aws:iam::0000000000:role/ExampleReadOnlyAccess`.

1. Now you can use the `tsh aws` command like the native `aws` command-line
   tool:

   ```code
   $ tsh aws s3 ls
   ```

1. Run the following command to log out of the AWS application and remove
   credentials:

   ```code
   $ tsh apps logout awsconsole
   ```

### Access applications using AWS SDKs

> 1. Log into the AWS Management Console app you configured earlier:

Reviewer 2: hrmm, i think either remove the `configured earlier` or perhaps (we can also probably word it similarly in the comment above this one)
```suggestion
1. Log into the AWS Management Console app you configured from this guide:
```

   ```code
   $ tsh apps login --aws-role ExampleReadOnlyAccess awsconsole
   ```

1. Use the following command to start a local HTTPS proxy server, which forwards
   AWS API traffic to the Teleport Application Service. Leave the terminal where

>    you run the command open since it runs in the foreground:

Reviewer 2: ```suggestion
   you ran the command open since it runs in the foreground:
```

Author: Suggested a clearer wording since "run" and "ran" can be ambiguous in this paragraph.

![CloudTrail](../../../img/application-access/cloud-trail.png)

## Troubleshooting

Read this section if you run into issues while following this guide.

### `Internal Server Error`

When visiting the AWS Management Console from the Teleport Web UI, you may see
an `InternalServer Error` message instead of the AWS Management Console. 

If this happens, check the Teleport Application Service logs:

<Tabs>
<TabItem label="EC2">

Execute the following command on the EC2 instance that runs the Teleport
Application Service:

```code
$ journalctl -u teleport
```

</TabItem>
<TabItem label="EKS">

```code
$ kubectl -n teleport-agent logs statefulset/teleport-kube-agent
```

</TabItem>
</Tabs>

If the Teleport Application Service encounters an error sending a request to the
AWS API, the logs will show the error message stack trace.

### The Application Service is not authorized to assume a role

> In the Teleport Application Service fails to assume the `ExampleReadOnlyAccess`

Reviewer 2: ```suggestion
If the Teleport Application Service fails to assume the `ExampleReadOnlyAccess`
```

## Next steps

> Now that you know how set up Teleport to protect access to the AWS Management

Reviewer 2: ```suggestion
Now that you know how to set up Teleport to protect access to the AWS Management
```
